openfisca_us/data/datasets/cps/cps.py

```diff
@@ -42,124 +42,122 @@ def generate(self, year: int):
+    cps["household_id"] = household.H_SEQ
+    cps["person_tax_unit_id"] = person.TAX_ID
+    cps["person_spm_unit_id"] = person.SPM_ID
+    cps["tax_unit_id"] = tax_unit.TAX_ID
+    cps["spm_unit_id"] = spm_unit.SPM_ID
+    cps["person_household_id"] = person.PH_SEQ
+    cps["person_family_id"] = person.PH_SEQ * 10 + person.PF_SEQ
+
+    # Add weights
+    # Weights are multiplied by 100 to avoid decimals
+    cps["person_weight"] = person.A_FNLWGT / 1e2
+    cps["family_weight"] = family.FSUP_WGT / 1e2
+
+    # Tax unit weight is the weight of the containing family.
+    family_weight = Series(
+        cps["family_weight"][...], index=cps["family_id"][...]
+    )
+    person_family_id = cps["person_family_id"][...]
+    persons_family_weight = Series(family_weight[person_family_id])
+    cps["tax_unit_weight"] = persons_family_weight.groupby(
+        cps["person_tax_unit_id"][...]
+    ).first()
+
+    cps["spm_unit_weight"] = spm_unit.SPM_WEIGHT / 1e2
+
+    cps["household_weight"] = household.HSUP_WGT / 1e2
+
+
+def add_personal_variables(cps: h5py.File, person: DataFrame):
+    """Add personal demographic variables.
+
+    Args:
+        cps (h5py.File): The CPS dataset file.
+        person (DataFrame): The CPS person table.
+    """
+    cps["age"] = np.where(
+        person.A_AGE.between(80, 85),
+        80 + 5 * np.random.rand(len(person)),
+        person.A_AGE,
+    )
+
+
+def add_personal_income_variables(cps: h5py.File, person: DataFrame):
+    """Add income variables.
+
+    Args:
+        cps (h5py.File): The CPS dataset file.
+        person (DataFrame): The CPS person table.
+    """
+    cps["employment_income"] = person.WSAL_VAL
+    cps["self_employment_income"] = person.SEMP_VAL
+    cps["e02100"] = person.FRSE_VAL
+    cps["social_security"] = person.SS_VAL
+    cps["e02300"] = person.UC_VAL
+
+    # Pensions/annuities
+    other_inc_type = person.OI_OFF
+    cps["e01500"] = other_inc_type.isin((2, 13)) * person.OI_VAL
+
+    # Alimony
+    cps["e00800"] = (person.OI_OFF == 20) * person.OI_VAL
+
+
+def add_SPM_variables(cps: h5py.File, spm_unit: DataFrame):
+    SPM_RENAMES = dict(
+        spm_unit_total_income="SPM_TOTVAL",
+        snap="SPM_SNAPSUB",
+        spm_unit_capped_housing_subsidy="SPM_CAPHOUSESUB",
+        free_school_meals="SPM_SCHLUNCH",
+        spm_unit_energy_subsidy="SPM_ENGVAL",
+        spm_unit_wic="SPM_WICVAL",
+        spm_unit_fica="SPM_FICA",
+        spm_unit_federal_tax="SPM_FEDTAX",
+        spm_unit_state_tax="SPM_STTAX",
+        spm_unit_work_childcare_expenses="SPM_CAPWKCCXPNS",
+        spm_unit_medical_expenses="SPM_MEDXPNS",
+        spm_unit_spm_threshold="SPM_POVTHRESHOLD",
+        spm_unit_net_income_reported="SPM_RESOURCES",
+    )
+
+    for openfisca_variable, asec_variable in SPM_RENAMES.items():
+        cps[openfisca_variable] = spm_unit[asec_variable]
+
+    cps["reduced_price_school_meals"] = cps["free_school_meals"][...] * 0
```

Comment:
What does this line do?

I think we need an intermediate variable for free+reduced meals, which we then deconstruct based on values per child. This variable might also include paid meal subsidies, per [this Census document](https://www.census.gov/content/dam/Census/library/working-papers/2021/demo/sehsd-wp2021-20.pdf#page=3).

Reply:
So far just avoids the microsim breaking, I think we should leave it in and come back to it later